{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3319a162",
   "metadata": {},
   "outputs": [],
   "source": [
    "from libs.Functions_For_PCA import *\n",
    "\n",
    "#number of PCA components, changing this number will result in very unexected results\n",
    "load_hap_names(\"libs/HapV3_Name.tsv\")\n",
    "load_hap_positions(\"libs/HapV3_Position.csv\")\n",
    "load_ref(\"libs/NC_045512.2_covid.fasta\")\n",
    "\n",
    "define_working_directory(\".\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d1c10f9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def projectData(data,ipca_file):\n",
    "    \n",
    "    #load the ipca\n",
    "    ipca = pk.load(open(ipca_file,'rb'))\n",
    "    #print(\"explained variance : \\n\"+str(ipca.explained_variance_))\n",
    "  \n",
    "    batchsize=30000\n",
    "    batch_projections=[]\n",
    "    for i in range(round(data.shape[0]/batchsize+0.5)):\n",
    "        (start,end)=(i*batchsize,(i+1)*batchsize)\n",
    "        print(\"project \"+str(start)+\":\"+str(end), end = '')\n",
    "        batch_projections.append(ipca.transform((data)[start:end]))\n",
    "        print(\"->ok\")\n",
    "    full_projection=np.concatenate(batch_projections, axis=0)\n",
    "  \n",
    "    \n",
    "    return(full_projection,ipca.explained_variance_)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "281d5474",
   "metadata": {},
   "outputs": [],
   "source": [
    "wp=\"/lustre03/project/6009524/shared/covid-19/database/data/inter/2021_01_20_imputed_hapV3/dataformat/working_on_pca/\"\n",
    "\n",
    "#createPCA(working_directory+\"paper_final_Alpha.data.df_10_noref.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c61c2068",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(142675, 12987)\n",
      "(142675, 2)\n",
      "reduce 142675 unique sequences to 45269 unique sequences \n",
      "reduce 329683 sequences to 109684 sequences \n",
      "project 0:30000->ok\n",
      "project 30000:60000->ok\n"
     ]
    }
   ],
   "source": [
    "\n",
    "set_of_hap=[\"V\",\"VIII\",\"X\",\"XI\",\"XV\"]\n",
    "\n",
    "\n",
    "data,rowinfos,snp_nb=loadData_filterHap(wp+\"paper_final_bothwave.data.df_10_noref.pkl\",set_of_hap)\n",
    "(full_projection,expl_var) = projectData(data,wp+\"paper_final_bothwave.data.df_10_noref_ipca.pkl\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "138e76c9",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-31-7bf0bb986b95>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0mvar\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvarfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdelim_whitespace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m \u001b[0mvar\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvar\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mvar\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'POS'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m>\u001b[0m\u001b[0;36m55\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0mvar\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'POS'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m<\u001b[0m\u001b[0;36m29854\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     20\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0mn_sample\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvar\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"A\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvar\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"C\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvar\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"G\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvar\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"T\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvar\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\".\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lustre03/project/6005588/shared/group_env/python_3.8.2/lib/python3.8/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__nonzero__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1440\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mfinal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1441\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__nonzero__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1442\u001b[0;31m         raise ValueError(\n\u001b[0m\u001b[1;32m   1443\u001b[0m             \u001b[0;34mf\"The truth value of a {type(self).__name__} is ambiguous. \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1444\u001b[0m             \u001b[0;34m\"Use a.empty, a.bool(), a.item(), a.any() or a.all().\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all()."
     ]
    }
   ],
   "source": [
    "\n",
    "datafile=\"/home/poujolra/projects/rrg-hussinju/shared/covid-19/database/data/inter/2021_01_20_imputed_hapV3/dataformat/working_on_pca/paper_final_Alpha.data\"\n",
    "varfile=\"/home/poujolra/projects/rrg-hussinju/shared/covid-19/database/data/inter/2021_01_20_imputed_hapV3/dataformat/working_on_pca/paper_final_Alpha.var\"\n",
    "\n",
    "minAlt=10\n",
    "\n",
    "saved_snpnames=datafile+\".snpNames_\"+str(minAlt)+\"_noref.pkl\"\n",
    "saved_df=datafile+\".df_\"+str(minAlt)+\"_noref.pkl\"\n",
    "\n",
    "########################################\n",
    "########################################\n",
    "########  SELECTION OF SNPS ############\n",
    "########################################\n",
    "########################################\n",
    "\n",
    "def select_SNP(datafile,varfile,minAlt):\n",
    "    listnuc=[\"A\",\"C\",\"G\",\"T\"]\n",
    "\n",
    "    var=pd.read_csv(varfile, delim_whitespace=True)\n",
    "    var=var[var['POS']>55]\n",
    "    var=var[var['POS']<29854]\n",
    "    n_sample=int(var[1:2][\"A\"])+int(var[1:2][\"C\"])+int(var[1:2][\"G\"])+int(var[1:2][\"T\"])+int(var[1:2][\".\"])\n",
    "    too_much_missing=str(len(var[var['.']>int(n_sample/10)]))\n",
    "\n",
    "    print(too_much_missing+ \" sites where remove because of more than \"+str(n_sample/10)+\" missing data (10%)\")\n",
    "    var=var[var['.']<=int(n_sample/10)]\n",
    "\n",
    "    snp_names=[]\n",
    "    for ref in listnuc :\n",
    "        alt_listnuc=[a for a in listnuc if a!=ref]\n",
    "        for alt in alt_listnuc:\n",
    "            d=list(var.loc[(var['REF']==ref) & (var[alt]> minAlt) & (var[alt]+var[\".\"]!= n_sample)][\"POS\"])\n",
    "            snp_names=snp_names+[str(i)+\"_\"+ref+\"_\"+alt for i in d]\n",
    "    nb_snp=len(snp_names)\n",
    "    print(str(nb_snp)+\" SNP kept with at least \"+str(minAlt)+\" alternative allele\")\n",
    "    snp_names=sorted(snp_names, key=lambda i: int(i.split(\"_\")[0]))\n",
    "    snplist=[(int(i.split(\"_\")[0]),i.split(\"_\")[2]) for i in snp_names]\n",
    "    return snplist\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6f77490",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "set_of_hap=[\"V\",\"VIII\",\"X\",\"XI\",\"XV\"]\n",
    "\n",
    "\n",
    "data,rowinfos,snp_nb=loadData_filterHap(working_directory+\"paper_final_bothwave.data.df_10_noref.pkl\",set_of_hap)\n",
    "\n",
    "(full_projection,expl_var) = projectData(data,working_directory+\"paper_final_bothwave.data.df_10_noref_ipca.pkl\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "511928fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df=pd.DataFrame(data=full_projection, index=data.index, columns=[\"PC\"+str(i) for i in range(1,21)])\n",
    "df[\"Haplotype\"]=rowinfos[\"Haplotype\"]\n",
    "df.to_csv(r'PC_VIII_V_VIII_X_XI_XV.csv',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "198206fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "working_directory=\"/lustre03/project/6009524/shared/covid-19/database/data/inter/2021_01_20_imputed_hapV3/dataformat/working_on_pca/\"\n",
    "\n",
    "data,rowinfos,snp_nb=loadData(working_directory+\"paper_final_Alpha.data.df_10_noref.pkl\")\n",
    "#rowinfos[\"Haplotype\"]=gethapdata(data)\n",
    "(full_projection,expl_var) = projectData(data,working_directory+\"paper_final_Alpha.data.df_10_noref_ipca.pkl\")\n",
    "p=plt.hist(rowinfos['sampleNumber'], bins = 20,log=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "da18dd9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cannot retrieve hap : no information for ^313_\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-21-5729e7c6f002>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     30\u001b[0m     \u001b[0;32mreturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgethap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m  \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgethapseqdata\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdat\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     31\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 32\u001b[0;31m \u001b[0mgethapdata\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-21-5729e7c6f002>\u001b[0m in \u001b[0;36mgethapdata\u001b[0;34m(dat)\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[0;31m#Return a list of haplotype name from a matrix and a list of position\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     29\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mgethapdata\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdat\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 30\u001b[0;31m     \u001b[0;32mreturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgethap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m  \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgethapseqdata\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdat\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     31\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[0mgethapdata\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-21-5729e7c6f002>\u001b[0m in \u001b[0;36mgethapseqdata\u001b[0;34m(dat)\u001b[0m\n\u001b[1;32m     22\u001b[0m     \u001b[0ma\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mp\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mHap_posList\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 24\u001b[0;31m         \u001b[0ma\u001b[0m\u001b[0;34m+=\u001b[0m\u001b[0mgetposdata\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdat\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mHap_posList\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mp\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     25\u001b[0m     \u001b[0;32mreturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-21-5729e7c6f002>\u001b[0m in \u001b[0;36mgetposdata\u001b[0;34m(dat, pos)\u001b[0m\n\u001b[1;32m     16\u001b[0m     \u001b[0mcollist\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfilter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mregex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpp\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mcollist\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m:\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Cannot retrieve hap : no information for \"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mpp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 18\u001b[0;31m     \u001b[0;32mreturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mgetposfromlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mcollist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     19\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;31m#Return a list of haplotype sequence from a matrix and a list of position\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lustre03/project/6005588/shared/group_env/python_3.8.2/lib/python3.8/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, func, axis, raw, result_type, args, **kwds)\u001b[0m\n\u001b[1;32m   7766\u001b[0m             \u001b[0mkwds\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7767\u001b[0m         )\n\u001b[0;32m-> 7768\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   7769\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7770\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mapplymap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna_action\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mOptional\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mDataFrame\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lustre03/project/6005588/shared/group_env/python_3.8.2/lib/python3.8/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mget_result\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    183\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_raw\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    184\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 185\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_standard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    186\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    187\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mapply_empty_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lustre03/project/6005588/shared/group_env/python_3.8.2/lib/python3.8/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mapply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    274\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    275\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mapply_standard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 276\u001b[0;31m         \u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mres_index\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_series_generator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    277\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    278\u001b[0m         \u001b[0;31m# wrap results\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lustre03/project/6005588/shared/group_env/python_3.8.2/lib/python3.8/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mapply_series_generator\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    288\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseries_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    289\u001b[0m                 \u001b[0;31m# ignore SettingWithCopy here in case the user mutates\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 290\u001b[0;31m                 \u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    291\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mABCSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    292\u001b[0m                     \u001b[0;31m# If we have a view on v, we need to make a copy because\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-21-5729e7c6f002>\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(row)\u001b[0m\n\u001b[1;32m     16\u001b[0m     \u001b[0mcollist\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfilter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mregex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpp\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mcollist\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m:\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Cannot retrieve hap : no information for \"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mpp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 18\u001b[0;31m     \u001b[0;32mreturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mgetposfromlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mcollist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     19\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;31m#Return a list of haplotype sequence from a matrix and a list of position\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-21-5729e7c6f002>\u001b[0m in \u001b[0;36mgetposfromlist\u001b[0;34m(row, cn)\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mc\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcn\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mc\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;32mreturn\u001b[0m \u001b[0mc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mcn\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mgethap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;32mreturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mHap_dict\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"NA\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "def getposfromlist(row,cn):\n",
    "    for c in cn :\n",
    "        if row[c]: return c[-1]\n",
    "    return cn[0][-3]\n",
    "\n",
    "def gethap(s): return(Hap_dict.get(s,\"NA\"))\n",
    "\n",
    "#Return the reference state of a given position\n",
    "def getposref(pos):\n",
    "    return refseq[pos]\n",
    "\n",
    "#Return the state of a given position for a list of sequence (rows)\n",
    "#Compute the subset of column names with the given position\n",
    "def getposdata(dat,pos):\n",
    "    pp=\"^\"+str(pos)+\"_\"\n",
    "    collist=list(dat.filter(regex=pp,axis=1).columns)\n",
    "    if collist==[] : print(\"Cannot retrieve hap : no information for \"+pp)\n",
    "    return(dat.apply (lambda row: getposfromlist(row,collist), axis=1))\n",
    "\n",
    "#Return a list of haplotype sequence from a matrix and a list of position\n",
    "def gethapseqdata(dat):\n",
    "    a=\"\"\n",
    "    for p in range(len(Hap_posList)):\n",
    "        a+=getposdata(dat,Hap_posList[p])\n",
    "    return(a)\n",
    "\n",
    "\n",
    "#Return a list of haplotype name from a matrix and a list of position\n",
    "def gethapdata(dat):\n",
    "    return([gethap(i) for  i in gethapseqdata(dat)])\n",
    "\n",
    "gethapdata(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e3a1f45",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "nbvert=2\n",
    "nbhori=5\n",
    "fig, axs = plt.subplots(nbvert,nbhori,figsize=(10*nbhori, 10*nbvert))\n",
    "\n",
    "def makeLabelTitle(i):\n",
    "    return('PC '+str(i+1)+\" (\"+str(int(expl_var[i]*1000)/1000)+\"%)\")\n",
    "\n",
    "\n",
    "#layer_top=[(i in index_of_ulist) for i in rowinfos.index]\n",
    "#layer_bottom=[(i not in index_of_ulist) for i in rowinfos.index]\n",
    "colortoplot={}\n",
    "colortoplot[True]='black'\n",
    "colortoplot[False]='blue'\n",
    "boolcolname=\"Alpha\"\n",
    "\n",
    "transparency=0.5\n",
    "for i in range(0,nbvert*nbhori):\n",
    "    ax=axs[int(i/nbhori),i%nbhori]\n",
    "    x_comp_ind=i*2\n",
    "    y_comp_ind=i*2+1\n",
    "    ax.scatter(full_projection[:,x_comp_ind],\n",
    "               full_projection[:,y_comp_ind],\n",
    "               s=list(rowinfos['sampleNumber']*8),\n",
    "               c=data[\"23063_A_T\"].map(colortoplot),\n",
    "               alpha=transparency)\n",
    "    ax.set(xlabel=makeLabelTitle(x_comp_ind),\n",
    "           ylabel=makeLabelTitle(y_comp_ind))\n",
    "TITLE=boolcolname+\" SNP_Number=\"+str(snp_nb)\n",
    "fig.suptitle(TITLE, fontsize=32)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5ba81117",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df=pd.DataFrame(data=full_projection, index=data.index, columns=[\"PC\"+str(i) for i in range(1,21)])\n",
    "#df[\"Haplotype\"]=rowinfos[\"Haplotype\"]\n",
    "df.to_csv(r'PC_Alpha.csv',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4481988",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "iscanadian=[]\n",
    "trueornot=[]\n",
    "istotallycanadian=[]\n",
    "for c in rowinfos.index:\n",
    "    #print(c)\n",
    "    where=0\n",
    "    isok=0\n",
    "    for cc in c.split(\"|\"):\n",
    "        if colmetadata[colmetadata[0]==cc].iloc[0][2]==\"Canada\":\n",
    "            where+=1\n",
    "        if colmetadata[colmetadata[0]==cc].iloc[0][1]:\n",
    "            isok+=1\n",
    "    iscanadian.append(where/len(c.split(\"|\"))>0.5)\n",
    "    istotallycanadian.append(where/len(c.split(\"|\"))!=1)\n",
    "    trueornot.append(isok/len(c.split(\"|\"))!=0)\n",
    "rowinfos[\"iscanadian\"]=iscanadian\n",
    "rowinfos[\"istotallycanadian\"]=iscanadian\n",
    "rowinfos[\"trueornot\"]=trueornot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "502b767f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "nbvert=2\n",
    "nbhori=2\n",
    "fig, axs = plt.subplots(nbvert,nbhori,figsize=(10*nbhori, 10*nbvert))\n",
    "\n",
    "def makeLabelTitle(i):\n",
    "    return('PC '+str(i+1)+\" (\"+str(int(expl_var[i]*1000)/1000)+\"%)\")\n",
    "\n",
    "\n",
    "#layer_top=[(i in index_of_ulist) for i in rowinfos.index]\n",
    "#layer_bottom=[(i not in index_of_ulist) for i in rowinfos.index]\n",
    "colortoplot={}\n",
    "colortoplot[True]='red'\n",
    "colortoplot[False]='blue'\n",
    "boolcolname=\"istotallycanadian\"\n",
    "\n",
    "transparency=0.5\n",
    "for i in range(0,nbvert*nbhori):\n",
    "    ax=axs[int(i/nbhori),i%nbhori]\n",
    "    x_comp_ind=i*2\n",
    "    y_comp_ind=i*2+1\n",
    "    ax.scatter(full_projection[:,x_comp_ind],\n",
    "               full_projection[:,y_comp_ind],\n",
    "               s=list(rowinfos['sampleNumber']*8),\n",
    "               c=rowinfos[boolcolname].map(colortoplot),\n",
    "               alpha=transparency)\n",
    "    ax.set(xlabel=makeLabelTitle(x_comp_ind),\n",
    "           ylabel=makeLabelTitle(y_comp_ind))\n",
    "TITLE=boolcolname+\" SNP_Number=\"+str(snp_nb)\n",
    "fig.suptitle(TITLE, fontsize=32)\n",
    "\n",
    "\n",
    "#custom_legend_shape=[Line2D([0], [0],marker='o', color=i, lw=4) for i in colhap]\n",
    "    \n",
    "#fig.legend(custom_legend_shape,namehap, title=\"Haplotypes:\",bbox_to_anchor=(1,0.815))\n",
    "#TITLE=wave + \" Nmin=\"+str(nMin)+ \" sampleNumber=\"+str(rowinfos['sampleNumber'].sum())+\" SNP_Number=\"+str(snp_nb)\n",
    "#fig.suptitle(TITLE, fontsize=32)\n",
    "\n",
    "#imageName=TITLE.replace(\" \",\"_\").replace(\"=\",\"\")\n",
    "#plt.savefig(imageName+'.png')\n",
    "#plt.savefig(imageName+'.pdf')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2eeb64f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "working_directory=\"/lustre03/project/6009524/shared/covid-19/database/data/inter/2021_01_20_imputed_hapV3/dataformat/working_on_pca/\"\n",
    "\n",
    "createPCA(working_directory+\"paper_final_VIIIfriends.data.df_10_noref.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b77ca232",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "data,rowinfos,snp_nb=loadData(working_directory+\"paper_final_VIIIfriends.data.df_10_noref.pkl\")\n",
    "#rowinfos[\"Haplotype\"]=gethapdata(data)\n",
    "(full_projection,expl_var) = projectData(data,working_directory+\"paper_final_VIIIfriends.data.df_10_noref_ipca.pkl\")\n",
    "p=plt.hist(rowinfos['sampleNumber'], bins = 20,log=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b298b6d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#plot the sample number hitogram \n",
    "#p=plt.hist(rowinfos['sampleNumber'], bins = 20,log=True)\n",
    "\n",
    "#load the ipca\n",
    "ipca = pk.load(open(datafile+\"_ipca.pkl\",'rb'))\n",
    "print(\"explained variance : \\n\"+str(ipca.explained_variance_))\n",
    "\n",
    "#Select onnly the haplotypes we want to project\n",
    "filterhap=rowinfos['Haplotype'].isin(Haplotypes_set)\n",
    "print(\"reduce \"+str(data.shape[0])+\" unique sequences to \"+\n",
    "      str(data[filterhap].shape[0])+\" unique sequences \")\n",
    "print(\"reduce \"+str(sum(rowinfos['sampleNumber']))+\n",
    "      \" sequences to \"+str(sum(rowinfos[filterhap]['sampleNumber']))+\" sequences \")\n",
    "rowinfos=rowinfos[filterhap]\n",
    "\n",
    "#Project\n",
    "full_projection=projectOnPCA(data[filterhap],ipca)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13b42983",
   "metadata": {},
   "outputs": [],
   "source": [
    "working_directory=\"/lustre03/project/6009524/shared/covid-19/database/data/inter/2021_01_20_imputed_hapV3/dataformat/working_on_pca/\"\n",
    "listavailableIPCA(working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6f14bf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Here you choose carefully your pca\n",
    "\n",
    "wave=\"firstwave\"\n",
    "nMin=10\n",
    "\n",
    "\n",
    "XVII_first_haps=[\"I\",\"II\",\"III\",\"IV\",\"V\",\"VI\",\"VII\",\"VIII\",\"IX\",\"X\",\"XI\",\"XII\",\"XIII\",\"XIV\",\"XV\",\"XVI\",\"XVII\"]\n",
    "\n",
    "(full_projection,rowinfos,expl_var,snp_nb)=loadEverything(wave,nMin,XVII_first_haps)\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3cf7c54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Colors for PCA plot\n",
    "\n",
    "colhap=[\"#003FFF\", \"#28573F\",\"#FF0000\",\"#FF964F\", \"#FFCAE4\", \"#CB5590\", \"#A6D4FF\", \"#AD26FF\", \"#A9A9A9\", \"#B3F396\", \"#40E0D0\", \"#8F1C55\",\"#03C0A4\", \"#FFD700\",\"#000000\",\"#009900\",\"#E3B778\"]\n",
    "namehap=[\"I\",\"II\",\"III\",\"IV\",\"V\",\"VI\",\"VII\",\"VIII\",\"IX\",\"X\",\"XI\",\"XII\",\"XIII\",\"XIV\",\"XV\",\"XVI\",\"XVII\"]\n",
    "\n",
    "XVIIhaplotypes_colors=dict(zip(namehap,colhap))\n",
    "\n",
    "\n",
    "\n",
    "nbvert=5\n",
    "nbhori=2\n",
    "fig, axs = plt.subplots(nbvert,nbhori,figsize=(10*nbhori, 10*nbvert))\n",
    "\n",
    "def makeLabelTitle(i):\n",
    "    return('component '+str(i+1)+\" (\"+str(int(expl_var[i]*1000)/1000)+\"%)\")\n",
    "\n",
    "totallist=[\"I\",\"II\",\"III\",\"IV\",\"V\",\"VI\",\"VII\",\"VIII\",\"IX\",\"X\",\"XI\",\"XII\",\"XIII\",\"XIV\",\"XV\",\"XVI\",\"XVII\"]\n",
    "toplist=[\"VIII\",\"X\",\"V\",\"XI\",\"XV\"]\n",
    "bottomlist=[i for i in totallist if i not in toplist]\n",
    "layer_top=rowinfos['Haplotype'].isin(toplist)\n",
    "layer_bottom=rowinfos['Haplotype'].isin(bottomlist)\n",
    "\n",
    "#layer_top=[(i in index_of_ulist) for i in rowinfos.index]\n",
    "#layer_bottom=[(i not in index_of_ulist) for i in rowinfos.index]\n",
    "\n",
    "transparency=0.5\n",
    "for lay in [layer_bottom,layer_top]:\n",
    "    for i in range(0,nbvert*nbhori):\n",
    "        ax=axs[int(i/nbhori),i%nbhori]\n",
    "        x_comp_ind=i*2\n",
    "        y_comp_ind=i*2+1\n",
    "        ax.scatter(full_projection[lay][:,x_comp_ind],\n",
    "                   full_projection[lay][:,y_comp_ind],\n",
    "                   s=list(rowinfos[lay]['sampleNumber']*4),\n",
    "                   c=rowinfos[lay]['Haplotype'].map(XVIIhaplotypes_colors),\n",
    "                   alpha=transparency)\n",
    "        ax.set(xlabel=makeLabelTitle(x_comp_ind),\n",
    "               ylabel=makeLabelTitle(y_comp_ind))\n",
    "    #transparency+=0.8\n",
    "\n",
    "custom_legend_shape=[Line2D([0], [0],marker='o', color=i, lw=4) for i in colhap]\n",
    "    \n",
    "fig.legend(custom_legend_shape,namehap, title=\"Haplotypes:\",bbox_to_anchor=(1,0.815))\n",
    "TITLE=wave + \" Nmin=\"+str(nMin)+ \" sampleNumber=\"+str(rowinfos['sampleNumber'].sum())+\" SNP_Number=\"+str(snp_nb)\n",
    "fig.suptitle(TITLE, fontsize=32)\n",
    "\n",
    "imageName=TITLE.replace(\" \",\"_\").replace(\"=\",\"\")\n",
    "#plt.savefig(imageName+'.png')\n",
    "plt.savefig(imageName+'.pdf')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25288e23",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### This part is for UMAP\n",
    "import umap\n",
    "working_directory=\"/lustre03/project/6009524/shared/covid-19/database/data/inter/2021_01_20_imputed_hapV3/dataformat/working_on_pca/\"\n",
    "listavailableIPCA(working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55adfcf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "define_working_directory(\"/lustre03/project/6009524/shared/covid-19/database/data/inter/2021_01_20_imputed_hapV3/dataformat/working_on_pca/\")\n",
    "listavailableIPCA(working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcdf5837",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "wave=\"bothwave\"\n",
    "nMin=10\n",
    "set_of_hap=[\"V\",\"VIII\",\"X\",\"XI\",\"XV\"]\n",
    "(full_projection,rowinfos,expl_var,snp_nb) = loadEverything(wave,nMin, set_of_hap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3caee5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################\n",
    "#Figure for VIII friends\n",
    "#################\n",
    "\n",
    "wave=\"bothwave\"\n",
    "nMin=10\n",
    "set_of_hap=[\"V\",\"VIII\",\"X\",\"XI\",\"XV\"]\n",
    "(full_projection,rowinfos,expl_var,snp_nb) = loadEverything(wave,nMin, set_of_hap)\n",
    "\n",
    "\n",
    "\n",
    "P1_list=['EPI_ISL_833136', 'EPI_ISL_833137', 'EPI_ISL_833138',\n",
    "        'EPI_ISL_833139', 'EPI_ISL_833140', 'EPI_ISL_833167',\n",
    "        'EPI_ISL_833169', 'EPI_ISL_833170', 'EPI_ISL_833171',\n",
    "        'EPI_ISL_833172', 'EPI_ISL_833173', 'EPI_ISL_833174',\n",
    "        'EPI_ISL_804824', 'EPI_ISL_804823', 'EPI_ISL_792680',\n",
    "        'EPI_ISL_792681', 'EPI_ISL_792682', 'EPI_ISL_792683',\n",
    "        'EPI_ISL_833366', 'EPI_ISL_833249']\n",
    "\n",
    "B117_File=\"B.1.1.7.epilist\"\n",
    "B117_list = [x for x in open(B117_File).read().split('\\n') if x]\n",
    "\n",
    "for i in rowinfos.index:\n",
    "    isplit=i.split('|')\n",
    "    if   isplit[0] in P1_list  : rowinfos.loc[i,'Haplotype']=\"P.1\"\n",
    "    elif isplit[0] in B117_list: rowinfos.loc[i,'Haplotype']=\"B.1.1.7\"\n",
    "        \n",
    "\n",
    "colhap=[\"#FF0000\"  ,\"#FFCAE4\",\"#AD26FF\",\"#B3F396\", \"#40E0D0\",\"#000000\",\"#FFFF00\"]\n",
    "set_of_hap=[\"P.1\"   ,\"V\"      ,\"VIII\"   ,\"X\"      ,\"XI\"      ,\"XV\"     ,\"B.1.1.7\"]\n",
    "colors_to_plot=dict(zip(set_of_hap,colhap))\n",
    "name_of_hap=[\"XV (Gamma)\"   ,\"V\"      ,\"VIII\"   ,\"X\"      ,\"XI\"      ,\"XV\"     ,\"XV (Alpha)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acfd78ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "data = pd.DataFrame(full_projection, index=list(rowinfos.index))\n",
    "print(data.shape)\n",
    "data['sampleNumber']=[i.split(\"|\") for i in list(data.index)]\n",
    "data = data.explode('sampleNumber')\n",
    "data.index=data['sampleNumber']\n",
    "data.drop('sampleNumber', inplace=True, axis=1)\n",
    "print(data.shape)\n",
    "\n",
    "md=0.3\n",
    "nn=50\n",
    "fit_exploded = umap.UMAP(random_state=1981,n_neighbors=nn,min_dist=md)\n",
    "fit_exploded.fit_transform(data)\n",
    "rowlist=[i.split(\"|\")[0] for i in list(rowinfos.index)]\n",
    "firstofgroup=[i in rowlist for i in list(data.index)]\n",
    "ex_sw=fit_exploded.embedding_[firstofgroup,0]\n",
    "ey_sw=fit_exploded.embedding_[firstofgroup,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "411ebc1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "print(ex.shape)\n",
    "print(data.shape)\n",
    "print(rowinfos.shape)\n",
    "\n",
    "\n",
    "axs = plt.subplots(1,1,figsize=(10,10))\n",
    "#print(fit.embedding_[[i in [\"P.1\"] for i in list(filtered_rowinfos['Haplotype'])],0])\n",
    "#print(fit.embedding_[[i in [\"P.1\"] for i in list(filtered_rowinfos['Haplotype'])],1])\n",
    "\n",
    "#for f in [[i not in [\"P.1\"] for i in list(rowinfos['Haplotype'])],[i in [\"P.1\"] for i in list(rowinfos['Haplotype'])]]:\n",
    "for f in [[True]*len(rowinfos)]:\n",
    "    filtered_rowinfos=rowinfos[f]\n",
    "    plt.scatter(ex, ey,\n",
    "                s=filtered_rowinfos.sampleNumber*8,\n",
    "                c=filtered_rowinfos['Haplotype'].map(colors_to_plot),\n",
    "                marker='.',\n",
    "                alpha=0.8)\n",
    "TITLE=wave + \" Nmin=\"+str(nMin)+ \" sampleNumber=\"+str(rowinfos['sampleNumber'].sum())+\" SNP_Number=\"+str(snp_nb)\n",
    "#TITLE=wave + \"_min_dist_\"+str(md)+ \"_sampleNumber_\"+str(nn)\n",
    "plt.title(TITLE);\n",
    "\n",
    "custom_legend_shape=[Line2D([0],[0],marker='.', color=i, lw=4) for i in colhap]\n",
    "\n",
    "plt.legend(custom_legend_shape,name_of_hap, title=\"\",bbox_to_anchor=(1,0.815))\n",
    "#plt.savefig('UMAP_'+TITLE+'.pdf')\n",
    "#plt.savefig('UMAPVIII_'+TITLE+'.png')\n",
    "print(TITLE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdd0bac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### UMAP first wave\n",
    "\n",
    "wave=\"firstwave\"\n",
    "nMin=10\n",
    "\n",
    "\n",
    "set_of_hap=[\"I\",\"II\",\"III\",\"IV\",\"V\",\"VI\",\"VII\",\"VIII\",\"IX\",\"X\",\"XI\",\"XII\",\"XIII\",\"XIV\",\"XV\",\"XVI\",\"XVII\"]\n",
    "\n",
    "(full_projection,rowinfos,expl_var,snp_nb)=loadEverything(wave,nMin,set_of_hap)\n",
    "\n",
    "colhap=[\"#003FFF\", \"#28573F\",\"#FF0000\",\"#FF964F\", \"#FFCAE4\", \"#CB5590\", \"#A6D4FF\", \"#AD26FF\", \"#A9A9A9\", \"#B3F396\", \"#40E0D0\", \"#8F1C55\",\"#03C0A4\", \"#FFD700\",\"#000000\",\"#009900\",\"#E3B778\"]\n",
    "colors_to_plot=dict(zip(set_of_hap,colhap))\n",
    "name_of_hap=set_of_hap\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa644773",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "data = pd.DataFrame(full_projection, index=list(rowinfos.index))\n",
    "print(data.shape)\n",
    "data['sampleNumber']=[i.split(\"|\") for i in list(data.index)]\n",
    "data = data.explode('sampleNumber')\n",
    "data.index=data['sampleNumber']\n",
    "data.drop('sampleNumber', inplace=True, axis=1)\n",
    "print(data.shape)\n",
    "\n",
    "md=0.3\n",
    "nn=2\n",
    "fit_exploded_fw = umap.UMAP(random_state=1981,n_neighbors=nn,min_dist=md)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed0c15c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fit_exploded_fw.fit_transform(data)\n",
    "#rowlist=[i.split(\"|\")[0] for i in list(rowinfos.index)]\n",
    "#firstofgroup=[i in rowlist for i in list(data.index)]\n",
    "#ex_fw=fit_exploded.embedding_[firstofgroup,0]\n",
    "#ey_fw=fit_exploded.embedding_[firstofgroup,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1d32e81",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fit = fit.exploded\n",
    "\n",
    "\n",
    "axs = plt.subplots(1,1,figsize=(10,10))\n",
    "#print(fit.embedding_[[i in [\"P.1\"] for i in list(filtered_rowinfos['Haplotype'])],0])\n",
    "#print(fit.embedding_[[i in [\"P.1\"] for i in list(filtered_rowinfos['Haplotype'])],1])\n",
    "\n",
    "#for f in [[i not in [\"P.1\"] for i in list(rowinfos['Haplotype'])],[i in [\"P.1\"] for i in list(rowinfos['Haplotype'])]]:\n",
    "plt.scatter(ex_fw, ey_fw,\n",
    "            s=rowinfos.sampleNumber*8,\n",
    "            c=rowinfos['Haplotype'].map(colors_to_plot),\n",
    "            marker='.',\n",
    "            alpha=0.8)\n",
    "TITLE=wave + \" Nmin=\"+str(nMin)+ \" sampleNumber=\"+str(rowinfos['sampleNumber'].sum())+\" SNP_Number=\"+str(snp_nb)\n",
    "#TITLE=wave + \"_min_dist_\"+str(md)+ \"_sampleNumber_\"+str(nn)\n",
    "plt.title(TITLE);\n",
    "\n",
    "#name_of_hap=[\"XV (Gamma)\"   ,\"V\"      ,\"VIII\"   ,\"X\"      ,\"XI\"      ,\"XV\"     ,\"XV (Alpha)\"]\n",
    "name_of_hap=namehap\n",
    "custom_legend_shape=[Line2D([0],[0],marker='.', color=i, lw=4) for i in colhap]\n",
    "\n",
    "plt.legend(custom_legend_shape,name_of_hap, title=\"\",bbox_to_anchor=(1,0.815))\n",
    "#plt.savefig('UMAP_'+TITLE+'.pdf')\n",
    "#plt.savefig('UMAPVIII_'+TITLE+'.png')\n",
    "print(TITLE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c0907cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### UMAP second wave\n",
    "\n",
    "wave=\"secondwave\"\n",
    "nMin=10\n",
    "\n",
    "\n",
    "set_of_hap=[\"I\",\"II\",\"III\",\"IV\",\"V\",\"VI\",\"VII\",\"VIII\",\"IX\",\"X\",\"XI\",\"XII\",\"XIII\",\"XIV\",\"XV\",\"XVI\",\"XVII\"]\n",
    "\n",
    "(full_projection,rowinfos,expl_var,snp_nb)=loadEverything(wave,nMin,set_of_hap)\n",
    "\n",
    "colhap=[\"#003FFF\", \"#28573F\",\"#FF0000\",\"#FF964F\", \"#FFCAE4\", \"#CB5590\", \"#A6D4FF\", \"#AD26FF\", \"#A9A9A9\", \"#B3F396\", \"#40E0D0\", \"#8F1C55\",\"#03C0A4\", \"#FFD700\",\"#000000\",\"#009900\",\"#E3B778\"]\n",
    "colors_to_plot=dict(zip(set_of_hap,colhap))\n",
    "name_of_hap=set_of_hap\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d821caf",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "data = pd.DataFrame(full_projection, index=list(rowinfos.index))\n",
    "print(data.shape)\n",
    "data['sampleNumber']=[i.split(\"|\") for i in list(data.index)]\n",
    "data = data.explode('sampleNumber')\n",
    "data.index=data['sampleNumber']\n",
    "data.drop('sampleNumber', inplace=True, axis=1)\n",
    "print(data.shape)\n",
    "\n",
    "md=0.3\n",
    "nn=2\n",
    "fit_exploded_fw = umap.UMAP(random_state=1981,n_neighbors=nn,min_dist=md)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca15ef63",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fit_exploded_fw.fit_transform(data)\n",
    "#rowlist=[i.split(\"|\")[0] for i in list(rowinfos.index)]\n",
    "#firstofgroup=[i in rowlist for i in list(data.index)]\n",
    "#ex_fw=fit_exploded.embedding_[firstofgroup,0]\n",
    "#ey_fw=fit_exploded.embedding_[firstofgroup,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c10b69bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "data = pd.DataFrame(full_projection, index=list(rowinfos.index))\n",
    "print(data.shape)\n",
    "data['sampleNumber']=[i.split(\"|\") for i in list(data.index)]\n",
    "data = data.explode('sampleNumber')\n",
    "data.index=data['sampleNumber']\n",
    "data.drop('sampleNumber', inplace=True, axis=1)\n",
    "print(data.shape)\n",
    "md=0.3\n",
    "nn=2\n",
    "fit_exploded_test = umap.UMAP(random_state=1981,n_neighbors=nn,min_dist=md)\n",
    "fit_exploded_test.fit_transform(data)\n",
    "rowlist=[i.split(\"|\")[0] for i in list(rowinfos.index)]\n",
    "firstofgroup=[i in rowlist for i in list(data.index)]\n",
    "ex_test=fit_exploded_test.embedding_[firstofgroup,0]\n",
    "ey_test=fit_exploded_test.embedding_[firstofgroup,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "669985b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "smalldata=data[0:5][[0,1,2,3]]\n",
    "print(\" --- UMAP data\")\n",
    "print(smalldata)\n",
    "smalltest = umap.UMAP(n_neighbors=2)\n",
    "smalltest.fit_transform(smalldata)\n",
    "print(\"\")\n",
    "print(\" --- UMAP results\")\n",
    "print(smalltest.embedding_[:,:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc353993",
   "metadata": {},
   "outputs": [],
   "source": [
    "smalldata=[[0,1,4,3,1],[0,1,4,3,1],]\n",
    "print(\" --- UMAP data\")\n",
    "print(smalldata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bc16c45",
   "metadata": {},
   "outputs": [],
   "source": [
    "nnn=[len(i.split('|')) for i in list(rowinfos.index)]\n",
    "xxx=fit_exploded_test.embedding_[:,0]\n",
    "yyy=fit_exploded_test.embedding_[:,1]\n",
    "print(xxx.shape)\n",
    "print(sum([len(i.split('|')) for i in list(rowinfos.index)]))\n",
    "print(rowinfos.shape)\n",
    "print(\"========\")\n",
    "start=0\n",
    "for i in nnn[:2]:\n",
    "    if i>1:\n",
    "        print(i)\n",
    "        print(xxx[start:start+i][:10])\n",
    "        print(data[start:start+i][:2])\n",
    "    start+=i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28435b25",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "axs = plt.subplots(1,1,figsize=(10,10))\n",
    "\n",
    "plt.scatter(ex_test, ey_test,\n",
    "            s=rowinfos.sampleNumber*8,\n",
    "            c=rowinfos['Haplotype'].map(colors_to_plot),\n",
    "            marker='.',\n",
    "            alpha=0.8)\n",
    "TITLE=wave + \" Nmin=\"+str(nMin)+ \" sampleNumber=\"+str(rowinfos['sampleNumber'].sum())+\" SNP_Number=\"+str(snp_nb)\n",
    "plt.title(TITLE);\n",
    "\n",
    "custom_legend_shape=[Line2D([0],[0],marker='.', color=i, lw=4) for i in colhap]\n",
    "\n",
    "plt.legend(custom_legend_shape,name_of_hap, title=\"\",bbox_to_anchor=(1,0.815))\n",
    "#plt.savefig('UMAP_'+TITLE+'.pdf')\n",
    "#plt.savefig('UMAPVIII_'+TITLE+'.png')\n",
    "print(TITLE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4588612a",
   "metadata": {},
   "outputs": [],
   "source": [
    "md=0.3\n",
    "nn=2\n",
    "fit_exploded_test_noexp = umap.UMAP(random_state=1981,n_neighbors=nn,min_dist=md)\n",
    "fit_exploded_test_noexp.fit_transform(full_projection)\n",
    "ex_test_noexp=fit_exploded_test_noexp.embedding_[True,0]\n",
    "ey_test_noexp=fit_exploded_test_noexp.embedding_[True,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1386b12",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "axs = plt.subplots(1,1,figsize=(10,10))\n",
    "\n",
    "plt.scatter(fit_exploded_test_noexp.embedding_[:,0],\n",
    "            fit_exploded_test_noexp.embedding_[:,1],\n",
    "            s=rowinfos.sampleNumber*8,\n",
    "            c=rowinfos['Haplotype'].map(colors_to_plot),\n",
    "            marker='.',\n",
    "            alpha=0.8)\n",
    "TITLE=wave + \" Nmin=\"+str(nMin)+ \" sampleNumber=\"+str(rowinfos['sampleNumber'].sum())+\" SNP_Number=\"+str(snp_nb)\n",
    "plt.title(TITLE);\n",
    "\n",
    "custom_legend_shape=[Line2D([0],[0],marker='.', color=i, lw=4) for i in colhap]\n",
    "\n",
    "plt.legend(custom_legend_shape,name_of_hap, title=\"\",bbox_to_anchor=(1,0.815))\n",
    "#plt.savefig('UMAP_'+TITLE+'.pdf')\n",
    "#plt.savefig('UMAPVIII_'+TITLE+'.png')\n",
    "print(TITLE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60d44e59",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fit = fit.exploded\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "axs = plt.subplots(1,1,figsize=(10,10))\n",
    "#print(fit.embedding_[[i in [\"P.1\"] for i in list(filtered_rowinfos['Haplotype'])],0])\n",
    "#print(fit.embedding_[[i in [\"P.1\"] for i in list(filtered_rowinfos['Haplotype'])],1])\n",
    "\n",
    "#for f in [[i not in [\"P.1\"] for i in list(rowinfos['Haplotype'])],[i in [\"P.1\"] for i in list(rowinfos['Haplotype'])]]:\n",
    "for f in [[True]*len(rowinfos)]:\n",
    "    filtered_rowinfos=rowinfos[f]\n",
    "    plt.scatter(fit.embedding_[f,0], fit.embedding_[f,1],\n",
    "                s=filtered_rowinfos.sampleNumber*8,\n",
    "                c=filtered_rowinfos['Haplotype'].map(colors_to_plot),\n",
    "                marker='.',\n",
    "                alpha=0.8)\n",
    "TITLE=wave + \" Nmin=\"+str(nMin)+ \" sampleNumber=\"+str(rowinfos['sampleNumber'].sum())+\" SNP_Number=\"+str(snp_nb)\n",
    "#TITLE=wave + \"_min_dist_\"+str(md)+ \"_sampleNumber_\"+str(nn)\n",
    "plt.title(TITLE);\n",
    "\n",
    "#name_of_hap=[\"XV (Gamma)\"   ,\"V\"      ,\"VIII\"   ,\"X\"      ,\"XI\"      ,\"XV\"     ,\"XV (Alpha)\"]\n",
    "name_of_hap=namehap\n",
    "custom_legend_shape=[Line2D([0],[0],marker='.', color=i, lw=4) for i in colhap]\n",
    "\n",
    "plt.legend(custom_legend_shape,name_of_hap, title=\"\",bbox_to_anchor=(1,0.815))\n",
    "#plt.savefig('UMAP_'+TITLE+'.pdf')\n",
    "#plt.savefig('UMAPVIII_'+TITLE+'.png')\n",
    "print(TITLE)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
